src/cogent3/core/new_alignment.py

```diff
@@ -4360,6 +4392,67 @@ def has_terminal_stop(self, gc: Any = None, strict: bool = False) -> bool:
                 return True
         return False
 
+    def sample(
+        self,
+        *,
+        n: int = None,
+        with_replacement: bool = False,
+        motif_length: int = 1,
+        randint=numpy.random.randint,
+        permutation=numpy.random.permutation,
+    ):
+        """Returns random sample of positions from self, e.g. to bootstrap.
+
+        Parameters
+        ----------
+        n
+            number of positions to sample. If None, all positions are sampled.
+        with_replacement
+            if True, samples with replacement.
+        motif_length
+            number of positions to sample as a single motif.
+        randint
+            random number generator, default is numpy.randint
+        permutation
+            function to generate a random permutation of positions, default is
+            numpy.permutation
+
+        Notes
+        -----
+        By default (resampling all positions without replacement), generates
+        a permutation of the positions of the alignment.
+
+        Setting with_replacement to True and otherwise leaving parameters as
+        defaults generates a standard bootstrap resampling of the alignment.
+        """
+        # refactor: type hint for randint, permutation
```

Comment:
add refactor array here too. Given `array_pos` property, it will be efficient to generate random indices and use `numpy.take()` using that. In the case of motif_length != 1, the total number of positions is just `len(self) // motif_length`. Having produced that, those indices can be scaled back up, or the numpy array reshaped...